### ISSUE TITLE
Bot didn't buy for full amount

### ISSUE THREAD

**Issue by schuesslerf** [fj317/PumpBot#65]
I just participated in a pump and noticed the following issue:

I set the bot to use 95% of my binance wallet amount and in the bot, it showed it correctly as well (using $1600)... Now in order history, I can see, that the bot only bought for 0.004 BTC, which is $200...  In the end, I still won money, but it didn't went optimal.... I guess it was a problem with the order and it couldn't fill the complete order.. In the console, I didn't see any problem though.

In another bot, I saw it splits the order in 4 (60% + 11% + 11% + 11%) - would this be an option? Do you have an explanation, why that happened?

Also attached the 2 screenshots, so you can check, if you see something.

![image](https://user-images.githubusercontent.com/10630224/110248255-74f1cb00-7f70-11eb-8cdf-55e07533398c.png)

![image](https://user-images.githubusercontent.com/10630224/110248274-82a75080-7f70-11eb-8c72-abe96ece1d3c.png)

**Comment by fj317** [comment]
> I set the bot to use 95% of my binance wallet amount and in the bot, it showed it correctly as well (using $1600)... Now in order history, I can see, that the bot only bought for 0.004 BTC, which is $200... In the end, I still won money, but it didn't went optimal.... I guess it was a problem with the order and it couldn't fill the complete order.. In the console, I didn't see any problem though.

I have zero ideas why the bot would use $200 instead of the $1600. Would you mind sending your config.json values as it could be something in there. 

> In another bot, I saw it splits the order in 4 (60% + 11% + 11% + 11%) - would this be an option? 

This is possible but it would require a lot of hard coding and for me personally I don't have the time to complete it all at the moment. If someone else wants to implement something like that they're welcome to.

**Comment by tank568** [comment]
Hi I think there's two possibilities here one is that maybe it went past the buyLimit percentage so you got a couple of coins before it passes? and also for the developer is it possible for the bot to buy and use full amount regardless of buyLimit? E.g. high price coin BCH at $500 and I want to use $2000, if I set buyLimit 50% it will only buy maybe 2.5 BCH and leaving like almost $700 unused although we got a good price in at $500

**Comment by fj317** [comment]
> Hi I think there's two possibilities here one is that maybe it went past the buyLimit percentage so you got a couple of coins before it passes? 

Well the bot simply creates a limit order with the quantity specified by the user and the price specified by whatever the buyLimit value is multiplied by the market value of the coin. Its possible that the reason the bot only bought $200 was because of this, but then this is an issue with Binance and the pump you participate in, and not the bot.

> Is it possible for the bot to buy and use full amount regardless of buyLimit?

I don't really understand what you're asking here. The buylimit is the maximum you will pay so if you set it at 0.5 you'll create an order at 50% the market value of the coin for the same amount of coin so it will never be fulfilled by anyone.

**Comment by tank568** [comment]
I mean like just trying to find a way to possibly used up all the funds so no unused fund would be left out in a pump

**Comment by tank568** [comment]
Anyway I participated in the pump above as well and I went in with $2000 and only bought about $150 worth. This is my config
![image](https://user-images.githubusercontent.com/79243837/110250711-74851e80-7fc8-11eb-8d84-c7c6d42364ae.png)

**Comment by fj317** [comment]
> Anyway I participated in the pump above as well and I went in with $2000 and only bought about $150 worth. This is my config
> ![image](https://user-images.githubusercontent.com/79243837/110250711-74851e80-7fc8-11eb-8d84-c7c6d42364ae.png)

It seems you're using old code since the config.json should be slightly different to the one you sent.

**Comment by tank568** [comment]
I just copied the useful ones so it wont show everything cause I don't think the endpoint or quotedcoin affects any unless you wanna see those as well

**Comment by schuesslerf** [comment]
That's my config.json:

```
{
    "quotedCoin": "BTC",
    "buyLimit": 1.90,
    "percentOfWallet": 95,
    "manualQuoted": 0.00000,
    "profitMargin": 150,
    "stopLoss": 0.90,
    "endpoint": "default",




    "comment": "DO NOT EDIT ANYTHING BELOW THIS LINE",
    "currentVersion" : 1.4
}
```

Don't really know what happened, as I couldn't see the buy order for the correct price...

**Comment by tank568** [comment]
wow buyLimit 1.9 and it happens? damn this is one tricky one

**Comment by schuesslerf** [comment]
@tank568 yeah, those pumps happen really fast, so this was just to made sure my order gets executed... The pump itself made about 450%.. took 50%, but only for $200 instead of the $1600..

**Comment by tank568** [comment]
is it possible that maybe a sell order only has 22 coins so it just buy that and stop? e.g. sell order wall is (in coins) 22,55,140 and all 3 at different price under my buyLimit but it only executed and bought that first 22 coins and stopped? Can't really figured out what might be a problem here tho to why this happens

**Comment by fj317** [comment]
I think I'll add more logging so if this happens again we have a lot more data we can use to understand what happened with the order. 

> 
> 
> is it possible that maybe a sell order only has 22 coins so it just buy that and stop? e.g. sell order wall is (in coins) 22,55,140 and all 3 at different price under my buyLimit but it only executed and bought that first 22 coins and stopped? Can't really figured out what might be a problem here tho to why this happens

And this is possible I guess. I'm not completely up to date as to how Binance runs their limit orders but I believe you can have a partially filled order so its possible your order was only filled partially due to the huge rise in price which meant the rest of the order couldn't be completed. 
To solve this I guess you just have to enter the coin into the program as fast as possible, or set the buyLimit to a much higher amount (like @schuesslerf did).

**Comment by schuesslerf** [comment]
@fj317 yeah, I also think binance limits some of these orders or there are problems with limit buy orders when lots of people buy it... usually a market order works better then (so the full amount gets executed)... because for me, it didn't work as well, even with a buyLimit of 1.9.... also, my order shows as fully filled, so it can't really be a partially filled order... 

But yeah, I guess the main thing here, for further development, is to add more logging...

**Comment by schuesslerf** [comment]
@fj317 not sure if it's a good option to just buy at market price... would make it more propable for the full amount to be executed (as I read on different threads etc.)... also, it would make sure, the order is executed... 

as an example: setting the buy limit (like I did) to 1,9 makes relatively sure I get the coin, but I'm getting it for a really high price.. in that case, a market order would have propably saved me some money... 
on the other side, a.f.a.i.k are often a bit above the price you'd set for limit buy order...

considering all points, the market buy order might be a legit choice though :)

**Comment by tank568** [comment]
No you wont be getting a higher price, e.g. coin X $100 and buyLimit 1.5 it will try to buy at $150 but it will automatically get the best price at that time probably $100. and yes we should try market price although it could seriously hurt some newbies, we need warnings everywhere haha.
I found this that can buy with your amount of BTC you have but I'm trying to find out how do you get the amount of coin you bought so that I can set a sell order
```
# buy order
order = client.order_market_buy(
    # order = client.create_test_order(
    symbol=tradingPair,
    quoteOrderQty=BTCtoSell
    )
```

**Comment by schuesslerf** [comment]
didn't check on the api but propably its in the order variable after execution? otherwise you should be able to check in the order history, on the official api, there's GET /api/v3/allOrders (HMAC SHA256) for that (to get past orders), don't know about the python lib tbh..

**Comment by schuesslerf** [comment]
> > In another bot, I saw it splits the order in 4 (60% + 11% + 11% + 11%) - would this be an option?
> 
> This is possible but it would require a lot of hard coding and for me personally I don't have the time to complete it all at the moment. If someone else wants to implement something like that they're welcome to.

@fj317 Just thought about this and and am thinking if its really so much effort... I think it's not more then just copy pasting lines 187 to 190 3 times and changing the quantity from amountOfCoin to amountOfCoin*0.6, respectively *0.1 or sth like that, so theres just these orders executed after another... Also, I'd again wanna suggest propably doing a market order for buying... Often, the suggestion is buying with market order and selling with limit order...

Here's a quote from a thread from bitcointalk:

> On exchanges with the possibility for “market orders” (like Binance), open a market order for buying and put a limit order for selling. On any other exchanges just place limit orders. Why not using a market sell? Well, during a pump there are huge rate fluctuations and if you are unlucky, your market order could get filled right when it just went down.
> When you place your buy order using a limit buy, you need to put the price a significant amount higher, than the current rate. Otherwise your order won’t get filled. I recommend a multiplier of 1.3 up to 1.6. Example: XYZ Coin is at rate 1$, thus your buy order will be 1.6$. You are basically simulating a market order by doing that. Why do I need to do that? Let's assume you want to buy the coin and you see the rate is at 1$, you place your order at 1$, but by the time your 1$ buy order is placed, the rate will be already at 1.3$ or even higher during a pump. Hence your order won’t get filled. IMPORTANT: In the case your limit buy order could not be filled after 3 oder 4 seconds, cancel it right away, otherwise it will be filled during the dump!!!

Might help in understand that point better :) 

So I guess the perfect prodecure would be (at least from my point of view): First placing a market buy order for the fulll amount (with the full amount, it shouldn't be necessary to place multiple ones, that all as quick as possible ofc), then placing multiple limit sell orders for different percentages, like in #63.

**Comment by fj317** [comment]
> > > In another bot, I saw it splits the order in 4 (60% + 11% + 11% + 11%) - would this be an option?
> >
> >This is possible but it would require a lot of hard coding and for me personally I don't have the time to complete it all at the moment. If someone else wants to implement something like that they're welcome to.
>
> @fj317 Just thought about this and and am thinking if its really so much effort... I think it's not more then just copy pasting lines 187 to 190 3 times and changing the quantity from amountOfCoin to amountOfCoin*0.6, respectively *0.1 or sth like that, so theres just these orders executed after another...

The problem with this is that there is no way to quickly see if the sell orders are being reached or not. If the price only reaches one of the sell limits and then doesn't reach the others before dipping down then you'll never sell your coins and be left with coins. Therefore it requires a lot of logic programming which I can't be bothered to write. 
If you want to do as you say, and just manually setup 4 different orders for 25% each time then simply create additional sell orders and set the quantity to 25% and sell.

>Also, I'd again wanna suggest propably doing a market order for buying... Often, the suggestion is buying with market order and selling with limit order...

As for the buy order, you can simply change the lines yourself I've mentioned this before so might add a section in the readme. Follow #51 on how to do this.

**Comment by Arceus786** [comment]
Hello! I also participated yesterday but I sold a bit late and made a little loss. But should I use this bot. Are u guys satisfied with it and do u recommend it? I am new to crypto trading and I will start with a small amount (200 bucks). Can this bot make me money?

**Comment by fj317** [comment]
> Hello! I also participated yesterday but I sold a bit late and made a little loss. But should I use this bot. Are u guys satisfied with it and do u recommend it? I am new to crypto trading and I will start with a small amount (200 bucks). Can this bot make me money?

Pump & dumps are all scams - 9/10 you'll lose money but you might occasionally get lucky depending on the pump. I'd recommend not to use the bot if you're a newbie since you'll inevitably lose money.
Also closing the issue since its now off topic.
